sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming.state
+
+import java.util.UUID
+
+import scala.util.Random
+
+import org.apache.hadoop.conf.Configuration
+
+import org.apache.spark.sql.execution.streaming.state.StateStoreTestsHelper.newDir
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types._
+
+class StateSchemaCompatibilityCheckerSuite extends SharedSparkSession {
+
+  testQuietly("changing schema of state when restarting query") {
+    val opId = Random.nextInt(100000)
+    val partitionId = -1
+
+    val hadoopConf: Configuration = new Configuration()
+
+    def runSchemaChecker(
+        dir: String,
+        queryId: UUID,
+        newKeySchema: StructType,
+        newValueSchema: StructType): Unit = {
+      // in fact, Spark doesn't support online state schema change, so need to check
+      // schema only once for each running of JVM
+      val providerId = StateStoreProviderId(
+        StateStoreId(dir, opId, partitionId), queryId)
+
+      new StateSchemaCompatibilityChecker(providerId, hadoopConf)
+        .check(newKeySchema, newValueSchema)
+    }
+
+    def verifyException(
+        oldKeySchema: StructType,
+        oldValueSchema: StructType,
+        newKeySchema: StructType,
+        newValueSchema: StructType): Unit = {
+      val dir = newDir()
+      val queryId = UUID.randomUUID()
+      runSchemaChecker(dir, queryId, oldKeySchema, oldValueSchema)
+
+      val e = intercept[StateSchemaNotCompatible] {
+        runSchemaChecker(dir, queryId, newKeySchema, newValueSchema)
+      }
+
+      e.getMessage.contains("Provided schema doesn't match to the schema for existing state!")
+      e.getMessage.contains(newKeySchema.json)
+      e.getMessage.contains(newValueSchema.json)
+      e.getMessage.contains(oldKeySchema.json)
+      e.getMessage.contains(oldValueSchema.json)
+    }
+
+    def verifySuccess(
+        oldKeySchema: StructType,
+        oldValueSchema: StructType,
+        newKeySchema: StructType,
+        newValueSchema: StructType): Unit = {
+      val dir = newDir()
+      val queryId = UUID.randomUUID()
+      runSchemaChecker(dir, queryId, oldKeySchema, oldValueSchema)
+      runSchemaChecker(dir, queryId, newKeySchema, newValueSchema)
+    }
+
+    val keySchema = new StructType()
+      .add(StructField("key1", IntegerType, nullable = true))
+      .add(StructField("key2", StringType, nullable = true))
+
+    val valueSchema = new StructType()
+      .add(StructField("value1", IntegerType, nullable = true))
+      .add(StructField("value2", StringType, nullable = true))
+
+    // adding field should fail
+    val fieldAddedKeySchema = keySchema.add(StructField("newKey", IntegerType))
+    val fieldAddedValueSchema = valueSchema.add(StructField("newValue", IntegerType))
+    verifyException(keySchema, valueSchema, fieldAddedKeySchema, fieldAddedValueSchema)
```

Comment:
I think that the changes to both key and value should be tested separately. If I understand correctly, then a change to either one is incompatible and should result in an error, but testing both at the same time could result in the failure for key when value is not checked or vice versa.